File: src/main/java/org/cascadebot/cascadebot/messaging/Messaging.java
/*
 * Copyright (c) 2019 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.messaging;

import lombok.experimental.UtilityClass;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.MessageBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageChannel;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.internal.utils.Checks;
import org.cascadebot.cascadebot.CascadeBot;
import org.cascadebot.cascadebot.Constants;
import org.cascadebot.cascadebot.Environment;
import org.cascadebot.cascadebot.MDCException;
import org.cascadebot.cascadebot.UnicodeConstants;
import org.cascadebot.cascadebot.data.language.Language;
import org.cascadebot.cascadebot.data.language.Locale;
import org.cascadebot.cascadebot.data.managers.GuildDataManager;
import org.cascadebot.cascadebot.data.objects.GuildData;
import org.cascadebot.cascadebot.exceptions.DiscordPermissionException;
import org.cascadebot.cascadebot.metrics.Metrics;
import org.cascadebot.cascadebot.utils.FormatUtils;
import org.cascadebot.cascadebot.utils.PasteUtils;
import org.cascadebot.cascadebot.utils.buttons.Button;
import org.cascadebot.cascadebot.utils.buttons.ButtonGroup;
import org.cascadebot.cascadebot.utils.pagination.Page;
import org.cascadebot.cascadebot.utils.pagination.PageCache;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

@UtilityClass
public final class Messaging {

    public static CompletableFuture<Message> sendMessageTypeMessage(MessageChannel channel, MessageType type, String message, boolean embed) {
        Checks.notNull(channel, "channel");
        Metrics.INS.messagesSent.labels(type.name()).inc();
        if (embed) {
            return channel.sendMessage(MessagingObjects.getMessageTypeEmbedBuilder(type).setDescription(message).build()).submit();
        } else {
            return channel.sendMessage(MessagingObjects.getMessageTypeMessageBuilder(type).append(message).build()).submit();
        }
    }

    public static CompletableFuture<Message> sendMessageTypeEmbedMessage(MessageChannel channel, MessageType type, EmbedBuilder builder, boolean embed) {
        Checks.notNull(channel, "channel");
        if (embed) {
            return channel.sendMessage(builder.setColor(type.getColor()).build()).submit();
        } else {
            return channel.sendMessage(type.getEmoji() + " " + FormatUtils.formatEmbed(builder.build())).submit();
        }
    }

    public static CompletableFuture<Message> sendInfoMessage(MessageChannel channel, String message) {
        return sendInfoMessage(channel, message, true);
    }

    public static CompletableFuture<Message> sendInfoMessage(MessageChannel channel, String message, boolean embed) {
        return sendMessageTypeMessage(channel, MessageType.INFO, message, embed);
    }

    public static CompletableFuture<Message> sendInfoMessage(MessageChannel channel, EmbedBuilder builder) {
        return sendMessageTypeEmbedMessage(channel, MessageType.INFO, builder, true);
    }

    public static CompletableFuture<Message> sendInfoMessage(MessageChannel channel, EmbedBuilder builder, boolean embed) {
        return sendMessageTypeEmbedMessage(channel, MessageType.INFO, builder, embed);
    }

    public static CompletableFuture<Message> sendSuccessMessage(MessageChannel channel, String message) {
        return sendSuccessMessage(channel, message, true);
    }

    public static CompletableFuture<Message> sendSuccessMessage(MessageChannel channel, String message, boolean embed) {
        return sendMessageTypeMessage(channel, MessageType.SUCCESS, message, embed);
    }

    public static CompletableFuture<Message> sendSuccessMessage(MessageChannel channel, EmbedBuilder builder) {
        return sendMessageTypeEmbedMessage(channel, MessageType.SUCCESS, builder, true);
    }

    public static CompletableFuture<Message> sendSuccessMessage(MessageChannel channel, EmbedBuilder builder, boolean embed) {
        return sendMessageTypeEmbedMessage(channel, MessageType.SUCCESS, builder, embed);
    }

    public static CompletableFuture<Message> sendWarningMessage(MessageChannel channel, String message) {
        return sendWarningMessage(channel, message, true);
    }

    public static CompletableFuture<Message> sendWarningMessage(MessageChannel channel, String message, boolean embed) {
        return sendMessageTypeMessage(channel, MessageType.WARNING, message, embed);
    }

    public static CompletableFuture<Message> sendWarningMessage(MessageChannel channel, EmbedBuilder builder) {
        return sendMessageTypeEmbedMessage(channel, MessageType.WARNING, builder, true);
    }

    public static CompletableFuture<Message> sendWarningMessage(MessageChannel channel, EmbedBuilder builder, boolean embed) {
        return sendMessageTypeEmbedMessage(channel, MessageType.WARNING, builder, embed);
    }

    public static CompletableFuture<Message> sendModerationMessage(MessageChannel channel, String message) {
        return sendModerationMessage(channel, message, true);
    }

    public static CompletableFuture<Message> sendModerationMessage(MessageChannel channel, String message, boolean embed) {
        return sendMessageTypeMessage(channel, MessageType.MODERATION, message, embed);
    }

    public static CompletableFuture<Message> sendModerationMessage(MessageChannel channel, EmbedBuilder builder) {
        return sendMessageTypeEmbedMessage(channel, MessageType.MODERATION, builder, true);
    }

    public static CompletableFuture<Message> sendModerationMessage(MessageChannel channel, EmbedBuilder builder, boolean embed) {
        return sendMessageTypeEmbedMessage(channel, MessageType.MODERATION, builder, embed);
    }

    public static CompletableFuture<Message> sendDangerMessage(MessageChannel channel, String message) {
        return sendDangerMessage(channel, message, true);
    }

    public static CompletableFuture<Message> sendDangerMessage(MessageChannel channel, String message, boolean embed) {
        return sendMessageTypeMessage(channel, MessageType.DANGER, message, embed);
    }

    public static CompletableFuture<Message> sendDangerMessage(MessageChannel channel, EmbedBuilder builder) {
        return sendMessageTypeEmbedMessage(channel, MessageType.DANGER, builder, true);
    }

    public static CompletableFuture<Message> sendDangerMessage(MessageChannel channel, EmbedBuilder builder, boolean embed) {
        return sendMessageTypeEmbedMessage(channel, MessageType.DANGER, builder, embed);
    }

    public static CompletableFuture<Message> sendExceptionMessage(MessageChannel channel, String s, Throwable e) {
        Locale locale = channel instanceof TextChannel ? Language.getGuildLocale(((TextChannel) channel).getGuild().getIdLong()) : Locale.getDefaultLocale();
        String message = Language.i18n(locale, "messaging.exception_message", s, PasteUtils.paste(PasteUtils.getStackTrace(MDCException.from(e))));
        if (Environment.isProduction()) {
            message += "\n" + Language.i18n(locale, "messaging.report_error", Constants.serverInvite);
        }
        return sendDangerMessage(channel, message);
    }

    public static void sendAutoDeleteMessage(MessageChannel channel, String message, long delay) {
        channel.sendMessage(message).queue(messageToDelete -> {
            // We should always be able to delete our own message
            messageToDelete.delete().queueAfter(delay, TimeUnit.MILLISECONDS);
        });
    }

    public static void sendAutoDeleteMessage(MessageChannel channel, MessageEmbed embed, long delay) {
        channel.sendMessage(embed).queue(messageToDelete -> {
            // We should always be able to delete our own message
            messageToDelete.delete().queueAfter(delay, TimeUnit.MILLISECONDS);
        });
    }

    public static void sendAutoDeleteMessage(MessageChannel channel, Message message, long delay) {
        channel.sendMessage(message).queue(messageToDelete -> {
            // We should always be able to delete our own message
            messageToDelete.delete().queueAfter(delay, TimeUnit.MILLISECONDS);
        });
    }

    public static CompletableFuture<Message> sendButtonedMessage(TextChannel channel, Message message, ButtonGroup buttonGroup) {
        Checks.notNull(message, "message");
        Checks.notNull(channel, "channel");
        Checks.notNull(buttonGroup, "button group");

        if (!channel.getGuild().getMember(CascadeBot.INS.getSelfUser()).hasPermission(channel, Permission.MESSAGE_ADD_REACTION)) {
            throw new DiscordPermissionException(Permission.MESSAGE_ADD_REACTION);
        }

        CompletableFuture<Message> future = channel.sendMessage(message).submit();
        future.thenAccept((sentMessage -> {
            buttonGroup.addButtonsToMessage(sentMessage);
            GuildDataManager.getGuildData(sentMessage.getGuild().getIdLong()).addButtonGroup(channel, sentMessage, buttonGroup);
        }));
        return future;
    }

    public static CompletableFuture<Message> sendButtonedMessage(TextChannel channel, String message, ButtonGroup buttonGroup) {
        Checks.notBlank(message, "message");
        return sendButtonedMessage(channel, new MessageBuilder().append(message).build(), buttonGroup);
    }

    public static CompletableFuture<Message> sendButtonedMessage(TextChannel channel, MessageEmbed embed, ButtonGroup buttonGroup) {
        Checks.notNull(embed, "embed");
        return sendButtonedMessage(channel, new MessageBuilder().setEmbed(embed).build(), buttonGroup);
    }

    public static CompletableFuture<Message> sendPagedMessage(TextChannel channel, Member owner, List<Page> pages) {
        ButtonGroup group = new ButtonGroup(owner.getIdLong(), channel.getIdLong(), channel.getGuild().getIdLong());
        group.addButton(new Button.UnicodeButton(UnicodeConstants.REWIND, (runner, textChannel, message) -> {
            PageCache.Pages pageGroup = GuildDataManager.getGuildData(textChannel.getGuild().getIdLong()).getPageCache().get(message.getIdLong());
            pageGroup.getPage(1).pageShow(message, 1, pageGroup.getPages());
            pageGroup.setCurrentPage(1);
        }));
        group.addButton(new Button.UnicodeButton(UnicodeConstants.BACKWARD_ARROW, (runner, textChannel, message) -> {
            PageCache.Pages pageGroup = GuildDataManager.getGuildData(textChannel.getGuild().getIdLong()).getPageCache().get(message.getIdLong());
            int newPage = pageGroup.getCurrentPage() - 1;
            if (newPage < 1) {
                return;
            }
            pageGroup.getPage(newPage).pageShow(message, newPage, pageGroup.getPages());
            pageGroup.setCurrentPage(newPage);
        }));
        group.addButton(new Button.UnicodeButton(UnicodeConstants.FORWARD_ARROW, (runner, textChannel, message) -> {
            PageCache.Pages pageGroup = GuildDataManager.getGuildData(textChannel.getGuild().getIdLong()).getPageCache().get(message.getIdLong());
            int newPage = pageGroup.getCurrentPage() + 1;
            if (newPage > pageGroup.getPages()) {
                return;
            }
            pageGroup.getPage(newPage).pageShow(message, newPage, pageGroup.getPages());
            pageGroup.setCurrentPage(newPage);
        }));
        group.addButton(new Button.UnicodeButton(UnicodeConstants.FAST_FORWARD, (runner, textChannel, message) -> {
            PageCache.Pages pageGroup = GuildDataManager.getGuildData(textChannel.getGuild().getIdLong()).getPageCache().get(message.getIdLong());
            pageGroup.getPage(pageGroup.getPages()).pageShow(message, pageGroup.getPages(), pageGroup.getPages());
            pageGroup.setCurrentPage(pageGroup.getPages());
        }));
        CompletableFuture<Message> future = channel.sendMessage(Language.i18n(channel.getGuild().getIdLong(), "messaging.loading_page")).submit();
        future.thenAccept(sentMessage -> {
            pages.get(0).pageShow(sentMessage, 1, pages.size());
            group.addButtonsToMessage(sentMessage);
            group.setMessage(sentMessage.getIdLong());
            GuildData guildData = GuildDataManager.getGuildData(channel.getGuild().getIdLong());
            guildData.addButtonGroup(channel, sentMessage, group);
            guildData.getPageCache().put(pages, sentMessage.getIdLong());
        });
        return future;
    }


}

File: src/main/kotlin/org/cascadebot/cascadebot/messaging/Messaging.kt
package org.cascadebot.cascadebot.messaging

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.MessageBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Emoji
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageChannel
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.interactions.components.ButtonStyle
import net.dv8tion.jda.internal.utils.Checks
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.Constants
import org.cascadebot.cascadebot.DiscordPermissionException
import org.cascadebot.cascadebot.Environment
import org.cascadebot.cascadebot.MDCException
import org.cascadebot.cascadebot.UnicodeConstants
import org.cascadebot.cascadebot.data.language.Language
import org.cascadebot.cascadebot.data.language.Locale
import org.cascadebot.cascadebot.data.managers.GuildDataManager
import org.cascadebot.cascadebot.metrics.Metrics
import org.cascadebot.cascadebot.utils.FormatUtils
import org.cascadebot.cascadebot.utils.PasteUtils
import org.cascadebot.cascadebot.utils.interactions.IButtonRunnable
import org.cascadebot.cascadebot.utils.interactions.ISelectionRunnable
import org.cascadebot.cascadebot.utils.interactions.CascadeActionRow
import org.cascadebot.cascadebot.utils.interactions.CascadeButton
import org.cascadebot.cascadebot.utils.interactions.CascadeSelectBox
import org.cascadebot.cascadebot.utils.interactions.ComponentContainer
import org.cascadebot.cascadebot.utils.interactions.InteractionMessage
import org.cascadebot.cascadebot.utils.pagination.Page
import org.cascadebot.cascadebot.utils.pagination.PageCache
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit

object Messaging {

    @JvmStatic
    @JvmOverloads
    fun sendMessage(type: MessageType, channel: MessageChannel, message: String, embed: Boolean = true): CompletableFuture<Message> {
        Metrics.INS.messagesSent.labels(type.name).inc()
        return if (embed) {
            channel.sendMessage(MessagingObjects.getMessageTypeEmbedBuilder(type).setDescription(message).build()).submit()
        } else {
            channel.sendMessage(MessagingObjects.getMessageTypeMessageBuilder(type)
                    .append(message)
                    .denyMentions(Message.MentionType.EVERYONE, Message.MentionType.HERE, Message.MentionType.ROLE).build())
                    .submit()
        }
    }

    @JvmStatic
    @JvmOverloads
    fun sendEmbedMessage(type: MessageType, channel: MessageChannel, builder: EmbedBuilder, embed: Boolean = true): CompletableFuture<Message> {
        return if (embed) {
            channel.sendMessage(builder.setColor(type.color).build()).submit()
        } else {
            channel.sendMessage(MessageBuilder()
                    .denyMentions(Message.MentionType.EVERYONE, Message.MentionType.HERE, Message.MentionType.ROLE)
                    .setContent(type.emoji + " " + FormatUtils.formatEmbed(builder.build()))
                    .build()).submit()
        }
    }

    @JvmStatic
    fun sendExceptionMessage(channel: MessageChannel, s: String, e: Throwable): CompletableFuture<Message> {
        val locale = if (channel is TextChannel) Language.getGuildLocale(channel.guild.idLong) else Locale.getDefaultLocale()
        var message = Language.i18n(locale, "messaging.exception_message", s, PasteUtils.paste(PasteUtils.getStackTrace(MDCException.from(e))))
        if (Environment.isProduction()) {
            message += """
                
                ${Language.i18n(locale, "messaging.report_error", Constants.serverInvite)}
                """.trimIndent()
        }
        return sendMessage(MessageType.DANGER, channel, message)
    }

    @JvmStatic
    fun sendAutoDeleteMessage(channel: MessageChannel, message: String, delay: Long) {
        channel.sendMessage(message).queue {
            // We should always be able to delete our own message
            it.delete().queueAfter(delay, TimeUnit.MILLISECONDS)
        }
    }

    @JvmStatic
    fun sendAutoDeleteMessage(channel: MessageChannel, embed: MessageEmbed, delay: Long) {
        channel.sendMessage(embed).queue {
            // We should always be able to delete our own message
            it.delete().queueAfter(delay, TimeUnit.MILLISECONDS)
        }
    }

    @JvmStatic
    fun sendAutoDeleteMessage(channel: MessageChannel, message: Message, delay: Long) {
        channel.sendMessage(message).queue {
            // We should always be able to delete our own message
            it.delete().queueAfter(delay, TimeUnit.MILLISECONDS)
        }
    }

    @JvmStatic
    fun sendComponentMessage(channel: TextChannel, message: String, container: ComponentContainer): CompletableFuture<Message> {
        return sendComponentMessage(channel, MessageBuilder().append(message).build(), container)
    }

    @JvmStatic
    fun sendComponentMessage(channel: TextChannel, embed: MessageEmbed, container: ComponentContainer): CompletableFuture<Message> {
        return sendComponentMessage(channel, MessageBuilder().setEmbeds(embed).build(), container)
    }

    @JvmStatic
    fun sendComponentMessage(channel: TextChannel, message: Message, container: ComponentContainer): CompletableFuture<Message> {
        val future = channel.sendMessage(message).setActionRows(container.getComponents().map { it.toDiscordActionRow() }).submit()
        future.thenAccept {
            GuildDataManager.getGuildData(it.guild.idLong).addComponents(channel, it, container)
        }

        return future
    }

    private val firstPageButton = CascadeButton.secondary(Emoji.fromUnicode(UnicodeConstants.REWIND),
        IButtonRunnable { _, textChannel, message ->
            val pageGroup = GuildDataManager.getGuildData(textChannel.guild.idLong).pageCache[message.idLong]
            handlePage(message, 1, pageGroup!!)
        })

    private val prevPageButton = CascadeButton.secondary("Prev Page", Emoji.fromUnicode(UnicodeConstants.BACKWARD_ARROW),
        IButtonRunnable { _, textChannel, message ->
            val pageGroup = GuildDataManager.getGuildData(textChannel.guild.idLong).pageCache[message.idLong]
            val newPage = pageGroup!!.currentPage - 1
            handlePage(message, newPage, pageGroup)
        })

    private val nextPageButton = CascadeButton.secondary("Next Page", Emoji.fromUnicode(UnicodeConstants.FORWARD_ARROW),
        IButtonRunnable { _, textChannel, message ->
            val pageGroup = GuildDataManager.getGuildData(textChannel.guild.idLong).pageCache[message.idLong]
            val newPage = pageGroup!!.currentPage + 1
            handlePage(message, newPage, pageGroup)
        })

    private val lastPageButton = CascadeButton.secondary(Emoji.fromUnicode(UnicodeConstants.FAST_FORWARD),
        IButtonRunnable { _, textChannel, message ->
            val pageGroup = GuildDataManager.getGuildData(textChannel.guild.idLong).pageCache[message.idLong]
            handlePage(message, pageGroup!!.pageCount, pageGroup)
        })

    private fun handlePage(message: InteractionMessage, newPage: Int, pageGroup: PageCache.Pages) {
        if (newPage < 1 || newPage > pageGroup.pageCount) {
            return
        }

        firstPageButton.disabled = false
        prevPageButton.disabled = false
        nextPageButton.disabled = false
        lastPageButton.disabled = false

        if (newPage <= 1) {
            firstPageButton.disabled = true
            prevPageButton.disabled = true
        }
        if (newPage >= pageGroup.pageCount) {
            nextPageButton.disabled = true
            lastPageButton.disabled = true
        }

        if (pageGroup.selectBox != null) {
            pageGroup.selectBox.clearDefaults()
            pageGroup.selectBox.addDefault("Page $newPage")
        }

        pageGroup.getPage(newPage).pageShow(message, newPage, pageGroup.pageCount)
        pageGroup.currentPage = newPage
    }

    @JvmStatic
    fun sendPagedMessage(channel: TextChannel, owner: Member, pages: List<Page>): CompletableFuture<Message>? {
        require(pages.isNotEmpty()) { "The number of pages cannot be zero!" }
        if (pages.size == 1) {
            val future = channel.sendMessage(Language.i18n(channel.guild.idLong, "messaging.loading_page")).submit()
            future.thenAccept { pages[0].pageShow(InteractionMessage(it, ComponentContainer()), 1, pages.size) }
            return future
        }
        val actionRow = CascadeActionRow()
        if (pages.size > 2) {
            actionRow.addComponent(firstPageButton)
            firstPageButton.disabled = true
        }
        actionRow.addComponent(prevPageButton)
        prevPageButton.disabled = true
        actionRow.addComponent(nextPageButton)
        if (pages.size > 2) {
            actionRow.addComponent(lastPageButton)
        }
        val container = ComponentContainer()
        container.addRow(actionRow)
        var select: CascadeSelectBox? = null;
            if (pages.size > 2) {
            select = CascadeSelectBox("pages-select",
                ISelectionRunnable { member: Member, textChannel: TextChannel, message: InteractionMessage, selected: List<String> ->
                    val pageGroup = GuildDataManager.getGuildData(textChannel.guild.idLong).pageCache[message.idLong]
                    var page: Int = selected[0].split(" ")[1].toInt()
                    handlePage(message, page, pageGroup!!)
                })
            var i = 1;
            for (page in pages) {
                if (page.title != null) {
                    select.addOption("Page " + i++, page.title)
                } else {
                    select.addOption("Page " + i++)
                }
            }
            select.addDefault("Page 1")
            val selectRow = CascadeActionRow()
            selectRow.addComponent(select)
            container.addRow(selectRow)
        }
        val future = sendComponentMessage(channel, Language.i18n(channel.guild.idLong, "messaging.loading_page"), container)
        future.thenAccept { sentMessage: Message ->
            pages[0].pageShow(InteractionMessage(sentMessage, container), 1, pages.size)
            val guildData = GuildDataManager.getGuildData(channel.guild.idLong)
            guildData.pageCache.put(sentMessage.idLong, pages, select)
        }
        return future
    }

}